File: src/main/services/parser-file.ts
import type { ParserCacheEntry } from '../../types/main-process';
import type { ParserState } from '../../types/overlay';

import * as fs from 'fs';
import { loadDungeonDataByName } from './parser-dungeon';
import { splitLogLine, unquote } from './parser-state';

const READ_STREAM_CHUNK_SIZE = 1024 * 1024;
const RECENT_DUNGEONS_TO_SCAN = 2;
const TAIL_SCAN_INITIAL_SIZE = 4 * 1024 * 1024;
const TAIL_SCAN_MAX_SIZE = 64 * 1024 * 1024;

async function processFileRange(
  filePath: string,
  start: number,
  endExclusive: number,
  entry: ParserCacheEntry,
  processLine: (state: ParserState, line: string) => void,
): Promise<void> {
  const length = Math.max(0, endExclusive - start);
  if (!length) return;

  await new Promise<void>((resolve, reject) => {
    let leftover = entry.leftover || '';
    const stream = fs.createReadStream(filePath, {
      start,
      end: endExclusive - 1,
      encoding: 'utf8',
      highWaterMark: READ_STREAM_CHUNK_SIZE,
    });

    const flushChunk = (chunk: string): void => {
      const text = leftover + chunk;
      const lines = text.split(/\r?\n/);
      leftover = lines.pop() || '';
      for (const line of lines) processLine(entry.state, line);
    };

    stream.on('data', flushChunk);
    stream.on('error', reject);
    stream.on('end', () => {
      entry.leftover = leftover;
      resolve();
    });
  });
}

function getFileIdentity(stat: { dev?: number; ino?: number }): string {
  return `${stat.dev || 0}:${stat.ino || 0}`;
}

function getLineDungeonBoundaryKind(line: string): 'start' | 'zone' | null {
  if (!line || (!line.includes('|DUNGEON_START|') && !line.includes('|ZONE_CHANGE|'))) {
    return null;
  }

  const parts: string[] = splitLogLine(line);
  if (parts.length < 3) return null;

  const event = parts[1] || null;
  if (event === 'DUNGEON_START') return 'start';
  if (event !== 'ZONE_CHANGE') return null;

  return loadDungeonDataByName(unquote(parts[2])) ? 'zone' : null;
}

function chooseRecentDungeonParseOffset(boundaryOffsets: number[], fallbackZoneOffset: number | null): number {
  if (boundaryOffsets.length >= RECENT_DUNGEONS_TO_SCAN) {
    return boundaryOffsets[Math.max(0, boundaryOffsets.length - RECENT_DUNGEONS_TO_SCAN)];
  }
  if (boundaryOffsets.length > 0) {
    return boundaryOffsets[0];
  }
  return fallbackZoneOffset ?? 0;
}

function nextLineStartOffset(buffer: any, index: number): number {
  const newlineIndex = buffer.indexOf(0x0a, index);
  if (newlineIndex === -1) return 0;
  return newlineIndex + 1;
}

function findBoundaryOffsetsInBuffer(
  buffer: any,
  absoluteStart: number,
): { starts: number[]; fallbackZoneOffset: number | null } {
  const starts: number[] = [];
  let fallbackZoneOffset: number | null = null;
  let lineStart = 0;

  for (let i = 0; i <= buffer.length; i += 1) {
    const isLineEnd = i === buffer.length || buffer[i] === 0x0a;
    if (!isLineEnd) continue;

    let lineEnd = i;
    if (lineEnd > lineStart && buffer[lineEnd - 1] === 0x0d) {
      lineEnd -= 1;
    }

    if (lineEnd > lineStart) {
      const lineBuffer = buffer.subarray(lineStart, lineEnd);
      if (lineBuffer.includes('|DUNGEON_START|') || lineBuffer.includes('|ZONE_CHANGE|')) {
        const kind = getLineDungeonBoundaryKind(lineBuffer.toString('utf8'));
        if (kind === 'start') {
          starts.push(absoluteStart + lineStart);
        } else if (kind === 'zone' && fallbackZoneOffset == null) {
          fallbackZoneOffset = absoluteStart + lineStart;
        }
      }
    }

    lineStart = i + 1;
  }

  return { starts, fallbackZoneOffset };
}

async function findRecentDungeonParseOffset(filePath: string, fileSize: number): Promise<number> {
  if (!fileSize) return 0;

  const handle = await fs.promises.open(filePath, 'r');

  try {
    let scanSize = Math.min(TAIL_SCAN_INITIAL_SIZE, fileSize);

    while (true) {
      const windowStart = Math.max(0, fileSize - scanSize);
      const windowSize = fileSize - windowStart;

      const buffer = Buffer.allocUnsafe(windowSize);
      const { bytesRead } = await handle.read(buffer, 0, windowSize, windowStart);
      const chunk = buffer.subarray(0, bytesRead);

      const sliceStart = windowStart > 0 ? nextLineStartOffset(chunk, 0) : 0;
      const slicedChunk = sliceStart > 0 ? chunk.subarray(sliceStart) : chunk;
      const absoluteSliceStart = windowStart + sliceStart;

      const { starts, fallbackZoneOffset } = findBoundaryOffsetsInBuffer(slicedChunk, absoluteSliceStart);

      if (starts.length >= RECENT_DUNGEONS_TO_SCAN || windowStart === 0) {
        return chooseRecentDungeonParseOffset(starts, fallbackZoneOffset);
      }

      if (scanSize >= TAIL_SCAN_MAX_SIZE) {
        return chooseRecentDungeonParseOffset(starts, fallbackZoneOffset) || absoluteSliceStart;
      }

      scanSize = Math.min(fileSize, Math.max(scanSize * 2, TAIL_SCAN_INITIAL_SIZE));
    }
  } finally {
    await handle.close();
  }
}

export {
  findRecentDungeonParseOffset,
  getFileIdentity,
  processFileRange,
};

File: src-tauri/src/parser_file.rs
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::path::Path;

const RECENT_DUNGEONS_TO_SCAN: usize = 2;
const TAIL_SCAN_INITIAL_SIZE: u64 = 4 * 1024 * 1024;
const TAIL_SCAN_MAX_SIZE: u64 = 64 * 1024 * 1024;

pub fn find_recent_dungeon_parse_offset(file_path: &Path, file_size: u64) -> Result<u64, String> {
    if file_size == 0 {
        return Ok(0);
    }

    let mut file = File::open(file_path).map_err(|error| error.to_string())?;
    let mut scan_size = TAIL_SCAN_INITIAL_SIZE.min(file_size);

    loop {
        let window_start = file_size.saturating_sub(scan_size);
        let window_size = file_size - window_start;
        let mut buffer = vec![0_u8; window_size as usize];
        file.seek(SeekFrom::Start(window_start))
            .map_err(|error| error.to_string())?;
        file.read_exact(&mut buffer)
            .map_err(|error| error.to_string())?;

        let slice_start = if window_start > 0 {
            next_line_start_offset(&buffer, 0)
        } else {
            0
        };
        let absolute_slice_start = window_start + slice_start as u64;
        let sliced_buffer = &buffer[slice_start..];
        let starts = find_dungeon_start_offsets(sliced_buffer, absolute_slice_start);

        if starts.len() >= RECENT_DUNGEONS_TO_SCAN || window_start == 0 {
            return Ok(choose_recent_dungeon_parse_offset(&starts, 0));
        }

        if scan_size >= TAIL_SCAN_MAX_SIZE {
            return Ok(choose_recent_dungeon_parse_offset(
                &starts,
                absolute_slice_start,
            ));
        }

        scan_size = file_size.min((scan_size * 2).max(TAIL_SCAN_INITIAL_SIZE));
    }
}

fn next_line_start_offset(buffer: &[u8], index: usize) -> usize {
    buffer[index..]
        .iter()
        .position(|byte| *byte == b'\n')
        .map(|position| index + position + 1)
        .unwrap_or(0)
}

fn find_dungeon_start_offsets(buffer: &[u8], absolute_start: u64) -> Vec<u64> {
    let mut starts = Vec::new();
    let mut line_start = 0_usize;

    for index in 0..=buffer.len() {
        let is_line_end = index == buffer.len() || buffer[index] == b'\n';
        if !is_line_end {
            continue;
        }

        let mut line_end = index;
        if line_end > line_start && buffer[line_end - 1] == b'\r' {
            line_end -= 1;
        }

        if line_end > line_start {
            let line = &buffer[line_start..line_end];
            if line
                .windows(b"|DUNGEON_START|".len())
                .any(|window| window == b"|DUNGEON_START|")
            {
                starts.push(absolute_start + line_start as u64);
            }
        }

        line_start = index + 1;
    }

    starts
}

fn choose_recent_dungeon_parse_offset(starts: &[u64], fallback_offset: u64) -> u64 {
    if starts.len() >= RECENT_DUNGEONS_TO_SCAN {
        starts[starts.len() - RECENT_DUNGEONS_TO_SCAN]
    } else if let Some(first) = starts.first() {
        *first
    } else {
        fallback_offset
    }
}
